src/java/org/apache/cassandra/index/sai/memory/TrieMemtableIndex.java

```diff
@@ -406,26 +399,11 @@ public CloseableIterator<PrimaryKeyWithSortKey> orderResultsBy(QueryContext quer
     /**
      * Count document frequencies for each term using brute force
      */
-    private BM25Utils.DocStats computeDocumentFrequencies(QueryContext queryContext, List<ByteBuffer> queryTerms)
+    private BM25Utils.DocStats computeDocumentFrequencies(List<ByteBuffer> queryTerms, AbstractAnalyzer docAnalyzer)
     {
-        var termIterators = keyIteratorsPerTerm(queryContext, Bounds.unbounded(indexContext.getPartitioner()), queryTerms);
         var documentFrequencies = new HashMap<ByteBuffer, Long>();
-        for (int i = 0; i < queryTerms.size(); i++)
-        {
-            // KeyRangeIterator.getMaxKeys is not accurate enough, we have to count them
-            long keys = 0;
-            for (var it = termIterators.get(i); it.hasNext(); )
-            {
-                PrimaryKey pk = it.next();
-                Cell<?> cellForKey = getCellForKey(pk);
-                if (cellForKey == null)
-                    // skip deleted rows
-                    continue;
-                keys++;
-            }
-            documentFrequencies.put(queryTerms.get(i), keys);
-        }
         long docCount = 0;
+        long totalTermCount = 0;
```

Comment:
Since we're just counting total terms inserted, we can skip this costly step here entirely and keep track of the value in each `TrieMemoryIndex` shard where we would increment it once per term inserted into the trie (and similarly decrement it on updates that remove terms).

Reply:
I would like to address this in a separate PR as this PR is already complex enough. It's still good or even must to have for #13553.

Reply:
Created https://github.com/riptano/cndb/issues/13997